transformer_engine/pytorch/attention.py

```diff
@@ -380,32 +560,64 @@ def forward(
             query_layer, key_layer, value_layer = [x.transpose(0,1).contiguous()
                            for x in (query_layer, key_layer, value_layer)]
 
-        batch_size, seqlen = query_layer.shape[0], query_layer.shape[1]
+        batch_size, max_seqlen_q, max_seqlen_kv = (
+            query_layer.shape[0], query_layer.shape[1], key_layer.shape[1]
+        )
 
         # [b, sq, np, hn]
         query_layer, key_layer, value_layer = [
             x.view(x.shape[0] * x.shape[1], *x.shape[2:])
             for x in [query_layer, key_layer, value_layer]
         ]
 
-        max_seqlen = seqlen
-        cu_seqlens = torch.arange(
-            0,
-            (batch_size + 1) * seqlen,
-            step=seqlen,
-            dtype=torch.int32,
-            device=query_layer.device)
+        if self.attn_mask_type != 'padding':
```

Comment:
The arbitrary mask case isn't handled. Perhaps we should add an assert that the attention mask type is either `"no_mask"` or `"causal"`.

Reply:
That's correct it doesn't. But the check for masks is handled in `DotProductAttention` already, so by the time it reaches any specific backend, only supported masking types are present.